rust/uefi/stub/src/main.rs

```diff
@@ -72,7 +74,60 @@ fn main(handle: Handle, mut system_table: SystemTable<Boot>) -> Status {
     let status;
     // A list of dynamically assembled initrds, e.g. credential initrds or system extension
     // initrds.
-    let dynamic_initrds: Vec<Vec<u8>> = Vec::new();
+    let mut dynamic_initrds: Vec<Vec<u8>> = Vec::new();
+
+    {
+        // This is a block for doing filesystem operations once and for all, related to companion
+        // files, nothing can open the LoadedImage protocol here.
+        // Everything must use `filesystem`.
+        let mut companions = Vec::new();
+        let mut filesystem = uefi::fs::FileSystem::new(
+            system_table
+                .boot_services()
+                .get_image_file_system(system_table.boot_services().image_handle())
+                .expect("Failed to open the simple filesystem for this image; netboot?"),
+        );
+        let default_dropin_directory;
+
+        if let Some(loaded_image_path) = pe_in_memory.file_path() {
+            default_dropin_directory = get_default_dropin_directory(
+                system_table.boot_services(),
+                loaded_image_path,
+                &mut filesystem,
+            )
+            .expect("Failed to obtain the default drop-in directory");
+        } else {
+            default_dropin_directory = None;
+        }
+
+        // TODO: how to do the proper .as_ref()? Should I take AsRef in the call definition… ?
+        companions.append(
+            &mut discover_credentials(
+                &mut filesystem,
+                default_dropin_directory.as_ref().map(|x| x.as_ref()),
+            )
+            .expect("Failed to discover system credentials"),
```

Comment:
This function also panics a lot. Is it necessary and can't we propagate the errors instead?

Reply:
Fixed.